packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -318,7 +318,7 @@ function i(
 /**
  * [f]lattening
  */
-function f(items: any[]): any[] {
+function f(items: Readonly<any[]>): Readonly<any[]> {
```

Comment:
```suggestion
function f(items: Array<VNodes[] | VNodes>): VNodes {
```

Reply:
Maybe I misunderstand the flatten API, but it looks like it takes the children array, so shouldn't the input (at least) be read-only? https://github.com/salesforce/lwc/blob/438323a485f3e18245f1c934e8cf963c0fde795a/packages/%40lwc/template-compiler/src/codegen/index.ts#L155-L160

I mean, maybe today it never receives an empty array, and in practice only the empty array is shared, but it still seems like the children should be consistently read-only everywhere.

Reply:
Hm, well now that `VNodes` is inherently readonly, this makes a lot of sense. I think the most conservative would be this:

```js
function f(items: Readonly<Array<Readonly<Array<VNodes>> | VNodes>>): VNodes {
```

A bit verbose, though? I dunno.

Reply:
Yeah, it's quite ugly, but I don't have an alternative proposal.